pkg/operation/botanist/component/vpnshoot/vpnshoot.go

```diff
@@ -0,0 +1,310 @@
+						},
+					},
+					Spec: corev1.PodSpec{
+						AutomountServiceAccountToken: pointer.Bool(false),
+						ServiceAccountName:           serviceAccount.Name,
+						PriorityClassName:            "system-cluster-critical",
+						DNSPolicy:                    corev1.DNSDefault,
+						NodeSelector:                 map[string]string{v1beta1constants.LabelWorkerPoolSystemComponents: "true"},
+						Tolerations: []corev1.Toleration{{
+							Key:      "CriticalAddonsOnly",
+							Operator: corev1.TolerationOpExists,
+						}},
+						Containers: []corev1.Container{
+							{},
+						},
+					},
+				},
+			},
+		}
+
+		networkPolicy = &networkingv1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "gardener.cloud--allow-vpn",
+				Namespace: metav1.NamespaceSystem,
+				Annotations: map[string]string{
+					v1beta1constants.GardenerDescription: "Allows the VPN to communicate with shoot components and makes the VPN reachable from the seed.",
+				},
+			},
+			Spec: networkingv1.NetworkPolicySpec{
+				PodSelector: metav1.LabelSelector{
+					MatchLabels: map[string]string{LabelKey: LabelValue},
+				},
+				Egress:      []networkingv1.NetworkPolicyEgressRule{{}},
+				Ingress:     []networkingv1.NetworkPolicyIngressRule{{}},
+				PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeEgress, networkingv1.PolicyTypeIngress},
+			},
+		}
+
+		service = &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "vpn-shoot",
+				Namespace: metav1.NamespaceSystem,
+				Labels:    map[string]string{LabelKey: LabelValue},
+			},
+			Spec: corev1.ServiceSpec{
+				Selector: map[string]string{LabelKey: LabelValue},
+				Type:     corev1.ServiceTypeLoadBalancer,
+				Ports: []corev1.ServicePort{
+					{
+						Name:       "openvpn",
+						Port:       servicePort,
+						TargetPort: intstr.FromInt(int(containerPort)),
+						Protocol:   corev1.ProtocolTCP,
+					},
+				},
+			},
+		}
+		vpa *autoscalingv1beta2.VerticalPodAutoscaler
+	)
+
+	//TODO: Add VPA in addons
+	if v.values.VPAEnabled {
+		vpaUpdateMode := autoscalingv1beta2.UpdateModeAuto
+		vpa = &autoscalingv1beta2.VerticalPodAutoscaler{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "vpn-shoot",
+				Namespace: metav1.NamespaceSystem,
+			},
+			Spec: autoscalingv1beta2.VerticalPodAutoscalerSpec{
+				TargetRef: &autoscalingv1.CrossVersionObjectReference{
+					APIVersion: appsv1.SchemeGroupVersion.String(),
+					Kind:       "Deployment",
+					Name:       deploymentName,
+				},
+				UpdatePolicy: &autoscalingv1beta2.PodUpdatePolicy{
+					UpdateMode: &vpaUpdateMode,
+				},
+				ResourcePolicy: &autoscalingv1beta2.PodResourcePolicy{
+					ContainerPolicies: []autoscalingv1beta2.ContainerResourcePolicy{
+						{
+							ContainerName: autoscalingv1beta2.DefaultContainerResourcePolicy,
+							MinAllowed: corev1.ResourceList{
+								corev1.ResourceCPU:    resource.MustParse("50m"),
+								corev1.ResourceMemory: resource.MustParse("150Mi"),
```

Comment:
this field was not present in the chart, but I found it used for the metrics server component.
Is it needed for VPN-shoot?

Reply:
Nice catch, I think it makes sense to add `minAllowed` to prevent that VPA is scaling down `vpn-shoot` too much. However, The values seem to high. Probably you can use `10m` CPU and `10Mi` memory.